src/Application/ConfigResolver.php

```diff
@@ -209,4 +205,30 @@ private static function preparePreset(array $preset, array $config): array
 
         return $preset;
     }
+
+    private static function resolvePreset(string $testPreset, Composer $composer): string
+    {
+        if ($testPreset === '' || $testPreset === 'default') {
+            return self::DEFAULT_PRESET;
+        }
+
+        $preset = self::guessPresetClass($testPreset, $composer);
+
+        if (Configuration::isValidPreset($preset)) {
+            return $preset;
+        }
+
+        throw new InvalidPreset('A preset must implement the ' . Preset::class . ' interface');
+    }
+
+    private static function guessPresetClass(string $preset, Composer $composer): string
+    {
+        foreach (self::PRESETS as $presetClass) {
+            if ($presetClass::shouldBeApplied($composer) || $preset === $presetClass::getName()) {
```

Comment:
I just test this, when I specify a custom preset, but there is an existing preset that match with my composer, my custom preset is not applied. 

I think this need to be reworked to directly take Preset in config if it exist and it is valid. In case preset is not in config, we try to guess which one apply from composer. 
In case there is no apply from composer, apply Default preset. 

WDYT ?